tf_quant_finance/black_scholes/asian_prices.py

```diff
@@ -0,0 +1,218 @@
+                                     dtype: tf.DType = None,
+                                     name: str = None) -> types.RealTensor:
+  """Computes the Black Scholes price for a batch of geometric averaging,
+  discrete time sampled call or put options.
+
+  In Black-Scholes, the marginal distribution of the underlying at each sampling
+  date is lognormal. The product of a sequence of lognormal variables is also
+  lognormal so we can re-express these options as vanilla options with modified
+  parameters and use the vanilla pricer to price them.
+
+  TODO: support volatility termstructures
+  For now we provide a single volatility for each option. A simple extension of
+  Black-Scholes provides a termstructure of volatilities that obeys non-negative
+  forward variance. We could also use this methodology under that model.
+
+  #### Example
+
+  ```python
+    # Price a batch of 5 seasoned discrete geometric asian options.
+    volatilities = np.array([0.0001, 102.0, 2.0, 0.1, 0.4])
+    forwards = np.array([1.0, 2.0, 3.0, 4.0, 5.0])
+    # Strikes will automatically be broadcasted to shape [5].
+    strikes = np.array([3.0])
+    # Expiries will be broadcast to shape [5], i.e. each option has strike=3
+    # and expiry = 1.
+    expiries = 1.0
+    sampling_times = np.array([[0.5, 0.5, 0.5, 0.5, 0.5], [1.0, 1.0, 1.0, 1.0, 1.0]])
+    past_fixings = np.array([[1.0, 2.0, 3.0, 4.0, 5.0]])
+    computed_prices = tff.black_scholes.discrete_geometric_average_price(
+        volatilities=volatilities,
+        strikes=strikes,
+        expiries=expiries,
+        forwards=forwards,
+        sampling_times=sampling_times,
+        past_fixings=past_fixings)
+  # Expected print output of computed prices:
+  # [ 0.        , 0.        , 0.52833763, 0.99555802, 1.91452834]
+  ```
+
+  #### References:
+  [1] Haug, E. G., The Complete Guide to Option Pricing Formulas. McGraw-Hill.
+
+  Args:
+    volatilities: Real `Tensor` of any shape and dtype. The volatilities to
+      expiry of the options to price.
+    strikes: A real `Tensor` of the same dtype and compatible shape as
+      `volatilities`. The strikes of the options to be priced.
+    expiries: A real `Tensor` of same dtype and compatible shape as
+      `volatilities`. The expiry of each option. The units should be such that
+      `expiry * volatility**2` is dimensionless.
+    spots: A real `Tensor` of any shape that broadcasts to the shape of the
+      `volatilities`. The current spot price of the underlying. Either this
+      argument or the `forwards` (but not both) must be supplied.
+    forwards: A real `Tensor` of any shape that broadcasts to the shape of
+      `volatilities`. The forwards to maturity. Either this argument or the
+      `spots` must be supplied but both must not be supplied.
+    sampling_times: A real `Tensor` of same dtype as expiries, n'th column is
+      the sampling times of the n'th option
```

Comment:
please specify the shape of `sampling_times`

Reply:
Fixed